design-proposals/migration-target.md

```diff
@@ -0,0 +1,151 @@
+  - During a planned maintenance window, I'm planning to drain more than one node in a sequence, so I want to be sure that the VM is going to land on a node that is not going to be drained in a near future (needing then a second migration) and being not interested in cordoning it also for other pods
+  - I just added a new node and I want to validate it trying to live migrate a specific VM there
+> [!NOTE]
+> technically all of this can be already achieved manipulating the node affinity rules on the VM object, but as a cluster admin I want to keep a clear boundary between what is a long-lasting setting for a VM, defined by the VM owner, and what is single shot requirement for a one-off migration
+- As a VM owner I don't want to see my VM object getting amended by another user just for maintenance reasons
+
+## Repos
+- https://github.com/kubevirt/kubevirt
+
+# Design
+## Proposed design
+We are going to add a new optional `nodeName` string field on the `VirtualMachineInstanceMigration` object.
+We are not going to alter by any mean the `spec` stanza of the VM or the VMI objects so future migrations or the node placement after a restart of the VM are not going to be affected by a `nodeName` set on a specific `VirtualMachineInstanceMigration` object.
+If the target pod fails to be started, the `VirtualMachineInstanceMigration` object will be marked as failed as it can already happen today for other reasons.
+
+## How to propagate the named node to the target virt-launcher pod
+We have two alternative approaches to propagate it to the target pod that are going to determine what is going to happen when the named node in the migration request is conflicting with other constrains set by the VM owner on the VM.
+
+### A. directly setting `spec.nodeName` on the target virt-launcher pod (bypassing the k8s scheduler)
+If the `nodeName` field is not empty, the migration controller will explicitly set `nodeName` on the virt-launcher pod that is going to be used as the target endpoint for the live migration.
+If the `nodeName` field is not empty, the k8s scheduler will ignore the Pod that is going to be used as the target for the migration and the kubelet on the named node will directly try to place the Pod on that node.
+
+When a pod is going to be executed, the scheduler is going to check it and, according to available cluster resources, nodeselectors, weighted affinity and anti-affinity rules and so on,
+the scheduler is going to select a node and write its name on `spec.nodeName` on the pod object. At this point the kubelet on the named node will try to execute the Pod on that node.
+
+If `spec.nodeName` is already set on a pod object as in this approach, the scheduler is not going to be involved in the process since the pod is basically already scheduled for that node and only for tha named node and so the kubelet on that node will directly try to execute it there eventually failing.
```

Comment:
I think using `pod.spec.nodeName` is likely the most straightforward approach. This does introduce some new failure modes that might not be obvious to admins.

For example, today if a target pod is unschedulable due to lack of resources, the migration object will time out due to the pod being stuck in "pending". This information is feed back to admin as an k8s event associated with the migration object.

However, by setting the pod.spec.NodeName directly, we'd be bypassing the checks that ensure the required resources are available on the node (like the node having the "kvm" device available for instance), and the pod would likely get scheduled and immediately fail. I don't think we are currently bubbling up these types of errors to the migration object, so this could leave admins wondering why their migration failed.

I guess what I'm trying to get at here is, I like this approach, let's make sure the new failure modes get reported back on the migration object so the Admin has some sort of clue as to why a migration has failed.

Reply:
@davidvossel We already report the failure reason on the VMIM. This is part of the VMIM status.

`pod.spec.nodeName` entirely bypassed the scheduler making AAQ unusable as it relies on "pod scheduling readiness".

From my pov, bypassing the scheduler is a no go.

Reply:
@vladikr @davidvossel +1.

`spec.nodeName` is a horrible field that is not being removed from Kubernetes only due to backward compatibility and causes a lot of trouble. I agree that it should be considered as a no-go.